Firmware/Marlin_main.cpp

```diff
@@ -5674,7 +5674,7 @@ if(eSoundMode!=e_SOUND_MODE_SILENT)
                 SERIAL_PROTOCOLPGM("\nZ search height: ");
                 SERIAL_PROTOCOL(MESH_HOME_Z_SEARCH);
                 SERIAL_PROTOCOLLNPGM("\nMeasured points:");
-                for (uint8_t y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
+                for (int16_t y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
```

Comment:
this can be written differently so that uint8_t is used, even though we are counting down.
```
for (uint8_t y = MESH_NUM_Y_POINTS; y-- > 0;) {
```

Reply:
Hmm perhaps. I will give it a try. Just for reference an array defined:

```c
float dummy_array[7][7] = {
                    {1, 2, 3, 4, 5, 6 ,7},
                    {8, 9, 10, 11, 12, 13 ,14},
                    {15, 16, 17, 18, 19, 20 ,21},
                    {22, 23, 24, 25, 26, 27 ,28},
                    {29, 30, 31, 32, 33, 34 ,35},
                    {36, 37, 38, 39, 40, 41 ,42},
                    {43, 44, 45, 46, 47, 48 ,49}
                };
```
Should return array items in this order after G81:
<img width="405" alt="image" src="https://user-images.githubusercontent.com/8218499/151668904-44428554-fdc7-42f4-9a52-e600c1678de9.png">

Reply:
I'm unable to optimise the loop. No matter what, using `int16_t` always yields lower flash memory. I wonder if this has something to do with memory alignment or if the compiler is making mistakes by trying to be too "smart".

Reply:
I disagree?
```
Before:
Sketch uses 232522 bytes (91%) of program storage space. Maximum is 253952 bytes.
Global variables use 6115 bytes of dynamic memory.

After:
Sketch uses 232514 bytes (91%) of program storage space. Maximum is 253952 bytes.
Global variables use 6115 bytes of dynamic memory.
```
![image](https://user-images.githubusercontent.com/17808203/151672917-5d607299-8275-4ec6-b3a7-7ee73cfff1d7.png)

Reply:
I think I understand why you couldn't get lower code size. When you used `for (uint8_t y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {`, the compiler threw a warning and optimized the code even more than it should be possible since, in theory, that was an endless loop. So it reduced the code size accordingly. Once you fix the behaviour, the code size increases a bit, but that should still be a bit lower than int16_t.

Reply:
Interesting I don't see a size reduction on MK3 branch either. I'm using Arduino IDE 1.8.19. I've set `LANG_MODE` to `0` in `config.h`

**Before:**
```
Sketch uses 231768 bytes (91%) of program storage space. Maximum is 253952 bytes.
Global variables use 6095 bytes of dynamic memory.
```
**After:**
```
Sketch uses 231782 bytes (91%) of program storage space. Maximum is 253952 bytes.
Global variables use 6095 bytes of dynamic memory.
```

Do you use 1.8.5?

Reply:
Yes. I'm using PF-build.sh which uses arduino 1.8.5 by default. I can make it use arduino 1.8.13, but newer arduino tends to be worse for this firmware. In fact, the most optimal code was achieved with 1.8.13. It's almost 2KB smaller than what we build right now. We might switch to that in the future, but it comes with the risk of uncovering bugs that didn't cause problems before. We should do that with caution. But in any case, we can start by optimizing the existing code. Some of the improvements carry over to the newer gcc, others are diminished, but in either case, smaller integer types all around help regardless of compiler version. What probably helps the most in the newer gcc is the better LTO which manages to optimize duplicate code better.

We should probably leave the various functions as they are now and focus more on the global variables and see which can be reduced. There are numerous global `long int` which are used for timekeeping in various features. These can be turned into ShortTimer in most cases (if max .expired(ms) is less than 65535ms).
Probably just as important is cleaning out stupid strings in ram. If you look at the .map file, you'll see how much junk there is. Just looking at a crash log and you'll see a bunch of messages in ram that shouldn't be there (should be in progmem).

Btw, I noticed an issue with the branch. At some point, something happened to it. If you look at the PR to MK3, there are many commits which seem to be unrelated to this PR. Any idea what happened? Some rebase issue? Accidentally merged something here that wasn't supposed to be merged? In any case, it's not looking too good. We might have to manually rebase the branch to the latest MK3 in order to solve these inconsistencies.
Also, this optimizations PR has quite a high priority right now. Soon we will start preparing a 3.12 branch. Yes, I know we didn't even release 3.11 yet, but 3.11 is not about the PRs that are open, it's just about the community languages and some bug fixes. 3.12 has tons of PRs, and we'll probably start checking and merging them from this early on in order to release it quicker (it will take a while to merge and test everything). We need as much flash as possible. I think the ram will suffice for now, but we can optimize it as well as that tends to save flash as well as we've seen so far.

So I think we should wrap things up with this smaller PR, fix the larger PR, merge it into 3.12, ..., profit!
Is there anything else to fix here besides this G81 issue?

Reply:
Yeah I think a rebase is needed. Regarding G81, I will change the loop index to uint8_t tomorrow and hopefully there will be no more changes from me :)

Perhaps this could help to sync this branch with upstream MK3 branch? I’ve used this to update my own fork: https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a

`git remote add upstream https://github.com/prusa3d/prusa-firmware.git`

`git fetch upstream`

`git rebase upstream/MK3`

`git push origin 1.11_Optimizations --force`

Then the history should be in sync.

Reply:
Now I am thinking if we can drop the commits manually with rebase? Its dangerous but should work. Its only the first 8 commits right? it looks to me like these commits were merged either manually onto the optimisation branch or came with MK3 but the history has been rewritten.

Reply:
![image](https://user-images.githubusercontent.com/17808203/151683590-aa86a4b7-1e9f-4a8e-89ea-4c215cbf9a9e.png)
There's more actually. Let's merge this and I'll try the rebase

Reply:
Wellll. I couldn't figure out how to rebase it cleanly. I can give you write access to my repo. Just please make sure to not overwrite the current branch in case something goes wrong. We can make a new PR to the MK3 from a different branch entirely

Reply:
Also, I just noticed that the branch name was 1.11... I'm not sure how I managed to name it like that :)

Reply:
I will create a new branch based on 1.11_Optimizations to test my rebase changes.